### PR TITLE
Add Inline Images and Inline Blocks API

### LINKED ISSUE #2043: Insert images inline
Users should be able to insert images inline in text.

Mockup: 

inserting images inline

It could work like this:

- You can only insert inline images if you use the global inserter
- You can only do it if you first set the cursor in text
- When you set the cursor in text, and click the global inserter, a blue square is immediately inserted where the inline element will appear
- The global inserter detects that the cursor is in text, and shows only inline blocks, if you'd like to insert a block level block just make a linebreak into an empty paragraph, or unfocus the paragraph
- Inline images have no placeholders, so as soon as you insert an inline image, it opens the media library

This flow is as close to the existing image insertion flows as I believe we can get. In the current editor, you always put the cursor where you want an image, first, and _then_ you insert an image. If this happens to be while you’re inside text, then you get an inline image, if you’re on an empty line, you get an image that behaves like a block level element. 

In this mockup, you'd have the same flow here, except with an Inserter UI instead of the usual Add Media way.

—

_Original text:_

If you're working within the Classic Text block, there's no way to include an image inline. Like the classic editor, we should support images within that block.

### PR DESCRIPTION
**Pull request by ellatrix**
## Description
This PR builds on #5794. Fixes #2043.

Summary, comparison to previous PR:

* Rebased with new inserter design.
* Moves inline token API under the `RichText` component.
* Move image specific logic to token registration.
* Uses a portal to add items in the inserter.
* Removes storing anything in the global application state. Everything goes through the portal.
* Insertion point is visible on hover. This is because the inserter has items with different insertion points. Also uses admin theme colour now.
* Inline items are visible together with block items when a `RichText` field has focus.
* Updates the e2e test. Deletes the image after the test is done.

## How has this been tested?
Place the caret in a `RichText` area. Open the inserter. "Inline image" should appear under "Inline blocks". On click the media modal opens. Select an image, and it should be inserted in the input field.

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. 
- [x] My code follows the accessibility standards. 
- [x] My code has proper inline documentation.

### CODE REVIEW

File: edit-post/index.js [hunk 1]
```diff
@@ -67,6 +68,7 @@ export function initializeEditor( id, post, settings ) {
 	const reboot = reinitializeEditor.bind( null, target, settings );
 
 	registerCoreBlocks();
+	registerCoreInlineBlocks();
```
Comment (iseulde) [inline edit-post/index.js]: I think we should just move these blocks in the `core-block` folder, even though the API would be slightly different.

Comment (mtias) [reply edit-post/index.js]: What about grouping with `RichText`?

Comment (iseulde) [reply edit-post/index.js]: Actually I was thinking of removing this separate folder and putting it in `core-blocks`, so it would all be registered together. These would have `inline` as the category. How does that sound?

Comment (mtias) [reply edit-post/index.js]: Mmm, not sure we should group these with blocks as they are different entities. To me, they are more like utilities for RichText, which is in turn a component.

File: core-inline-blocks/inline-image/index.js [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+
+	title: __( 'Inline Image' ),
+
+	type: 'image',
+
+	icon: 'format-image',
+
+	category: 'inline',
+
+	edit( { onSave } ) {
```
Comment (jayshenk) [inline core-inline-blocks/inline-image/index.js]: I like how you added `edit` and `save` functions to provide an interface similar to regular blocks. I wonder if it will be possible to have the edit toolbar in https://github.com/WordPress/gutenberg/pull/5794#issuecomment-381516924 be part of this `edit` function as well? Not sure if this is possible though, since I think the React element returned from the `save` function will be rendered to string before using the edit toolbar.

File: editor/components/rich-text/inline-blocks.js [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+		return (
+			<Fragment>
+				{ isInlineInsertionPointVisible &&
+					<div
+						style={ { position: 'absolute', ...this.getInsertPosition() } }
+						className="blocks-inline-insertion-point"
+					/>
+				}
+				{ type &&
+					<type.edit onSave={ this.onSave( type ) } />
```
Comment (jayshenk) [inline editor/components/rich-text/inline-blocks.js]: Nice how you moved the Media Library logic out of this component by defining an `edit` function for inline blocks and moving the `MediaUpload` component to the `edit` function of the inline image.

**Comment by jayshenk** [comment]
Cool how you have inline blocks working with the new Inserter design! Just noting there was previous discussion in https://github.com/WordPress/gutenberg/pull/5794#issuecomment-382684045 about how at this point we may want to also show regular blocks in the menu at the same time, and there is a new mockup that is being discussed in https://github.com/WordPress/gutenberg/issues/6998#issuecomment-392736034 for this. CC @jasmussen

File: core-inline-blocks/inline-image/index.js (hunk 2 above)
Comment (iseulde) [reply core-inline-blocks/inline-image/index.js]: Probably will have to be some different method. We can name these differently based on what it's for. Maybe insert, edit/select, and save or similar.

File: edit-post/index.js (hunk 1 above)
Comment (iseulde) [reply edit-post/index.js]: I moved some stuff around. Still polishing.

**Comment by iseulde** [comment]
@jasmussen Question: if we show both inline and block items in the inserter at the same time, how doe we show insertion points?

**Comment by iseulde** [comment]
Should we do it again on hover as it used to be?

File: components/slot-fill/slot.js [hunk 4]
```diff
@@ -54,13 +54,9 @@ class Slot extends Component {
 
 		const fills = map( getFills( name ), ( fill ) => {
 			const fillKey = fill.occurrence;
+			const fillChildren = isFunction( fill.props.children ) ? fill.props.children( fillProps ) : fill.props.children;
```
Comment (iseulde) [inline components/slot-fill/slot.js]: I made this change to allow `null` values.

**Comment by iseulde** [comment]
e2e tests fail: Multi-block selection and the new one. locally it passes. Trying to figure that one out haha

**Comment by iseulde** [comment]
E2e tests are passing locally... Unsure why they are not passing on Travis.

**Comment by aduth** [comment]
The tests fail for me locally. Are you using Docker for the testing?

If I had to guess based on the error message, the changes to the button click by `aria-label` are problematic (not finding the node).

File: editor/components/rich-text/tokens/ui.js [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+ */
+import './style.scss';
+import { InserterResultsPortal } from '../../inserter';
+import { getTokenSettings } from './registration';
+
+export default class TokenUI extends Component {
+	constructor() {
+		super( ...arguments );
+
+		this.getInsertPosition = this.getInsertPosition.bind( this );
```
Comment (mcsf) [inline editor/components/rich-text/tokens/ui.js]: Doesn't need to be bound, since it's always called with context.

File: editor/components/rich-text/index.js [hunk 6]
```diff
@@ -889,6 +891,12 @@ export class RichText extends Component {
 						{ formatToolbar }
 					</div>
 				) }
+				{ isSelected &&
+					<TokenUI
```
Comment (mcsf) [inline editor/components/rich-text/index.js]: Would a better name be `InsertableRichTextTokens`, `InsertableInlineTokens`?

Comment (mcsf) [reply editor/components/rich-text/index.js]: I'm also a bit torn on the module name `./tokens/ui`. When reviewing this diff it took me longer to distinguish the pieces. Would something more explicit like `./tokens/inserter-fill` be clearer?

File: editor/components/rich-text/tokens/ui.js [hunk 7]
```diff
@@ -0,0 +1,88 @@
...
+			this.setState( { selected: null } );
+		};
+	}
+
+	onHover( settings ) {
+		this.setState( { hovered: !! settings } );
+	}
+
+	onSelect( settings ) {
+		this.setState( { selected: settings } );
```
Comment (mcsf) [inline editor/components/rich-text/tokens/ui.js]: Super minor preference: `this.setState( { selected } );`

**Comment by iseulde** [comment]
@aduth No, I'm not using Docker. The nodes appear here. Is it a translation issue?

screen shot 2018-06-01 at 16 14 54

File: editor/components/rich-text/tokens/ui.js [hunk 8]
```diff
@@ -0,0 +1,88 @@
...
+	}
+
+	render() {
+		const { hovered, selected } = this.state;
+
+		return (
+			<Fragment>
+				<InserterResultsPortal
+					title={ __( 'Inline Rich Text Tokens' ) }
```
Comment (mcsf) [inline editor/components/rich-text/tokens/ui.js]: Did we decide against _Inline Blocks_? Regardless, feels like we can improve over this wordier and technical title.

**Comment by iseulde** [comment]
We need to insert the blocks without searching because when searching for "Image" both block and token will show up.

File: test/e2e/support/utils.js [hunk 9]
```diff
@@ -103,10 +103,7 @@ export async function insertBlock( searchTerm ) {
 	// Waiting here is necessary because sometimes the inserter takes more time to
 	// render than Puppeteer takes to complete the 'click' action
 	await page.waitForSelector( '.editor-inserter__menu' );
-	await page.keyboard.type( searchTerm );
-	await page.keyboard.press( 'Tab' );
-	await page.keyboard.press( 'Tab' );
-	await page.keyboard.press( 'Enter' );
+	await page.click( `button[aria-label="${ searchTerm }"]` );
```
Comment (mcsf) [inline test/e2e/support/utils.js]: Why this change? Won't it break with blocks that aren't featured (and hidden in the collapsed panels)?

Comment (iseulde) [reply test/e2e/support/utils.js]: See https://github.com/WordPress/gutenberg/pull/6959#issuecomment-393894222.

File: editor/components/rich-text/tokens/ui.js (hunk 8 above)
Comment (iseulde) [reply editor/components/rich-text/tokens/ui.js]: We want to avoid calling init blocks I think. Maybe just inline tokens? Cc @mtias.

File: test/e2e/support/utils.js (hunk 9 above)
Comment (youknowriad) [reply test/e2e/support/utils.js]: In my testing, this is the change that's causing the breakage in e2e tests. I understand the need to base the insertion on something less ambiguous than search. Maybe this should be a separate utility `insertBlockByLabel`

File: editor/components/rich-text/index.js (hunk 6 above)
Comment (iseulde) [reply editor/components/rich-text/index.js]: Yeah, makes sense.

File: components/slot-fill/slot.js (hunk 4 above)
Comment (iseulde) [reply components/slot-fill/slot.js]: How so? There's still a key added. This change is needed because you can't return null when using the function. See https://github.com/WordPress/gutenberg/pull/6959/files#diff-8c9cd11f122f43914bc769f4ad5c36d9R21.

File: test/e2e/support/utils.js (hunk 9 above)
Comment (iseulde) [reply test/e2e/support/utils.js]: Sure, but that separate utility will still fail? I'm unsure why `await page.click( `button[aria-label="${ searchTerm }"]` );` is failing elsewhere.

File: editor/components/rich-text/index.js (hunk 6 above)
Comment (iseulde) [reply editor/components/rich-text/index.js]: Actually `inserter-fill` is not the only thing it's doing. We can move it in a folder though.

File: editor/components/rich-text/tokens/ui.js (hunk 7 above)
Comment (iseulde) [reply editor/components/rich-text/tokens/ui.js]: There's also:

``js
onHover( settings ) {
	this.setState( { hovered: !! settings } );
}
```

File: editor/components/rich-text/tokens/ui.js (hunk 8 above)
Comment (mtias) [reply editor/components/rich-text/tokens/ui.js]: If it's exposed to users, maybe "inline elements".

**Comment by iseulde** [comment]
The error seems fixed, but now there is another. Locally everything still passes:

screen shot 2018-06-01 at 17 58 03

**Review by youknowriad — COMMENTED** [review]
Overall, this feels solid to me. 👏

File: editor/components/rich-text/core-tokens/image/index.js [hunk 10]
```diff
@@ -0,0 +1,45 @@
...
+export const settings = {
+	id: 'image',
+
+	title: __( 'Inline Image' ),
+
+	type: 'image',
+
+	icon: 'format-image',
+
+	category: 'inline',
```
[the remark this answers is not in the document]
Comment (iseulde) [reply editor/components/rich-text/core-tokens/image/index.js]: Good catch, this is a remnant from previous approaches.

File: editor/components/rich-text/index.js [hunk 11]
```diff
@@ -939,6 +947,8 @@ RichText.defaultProps = {
 	format: 'element',
 };
 
+registerCoreTokens();
```
[the remark this answers is not in the document]
Comment (iseulde) [reply editor/components/rich-text/index.js]: I feel like I discussed this with @mtias somewhere, but can't find the conversation now.

File: editor/components/rich-text/style.scss [hunk 12]
```diff
@@ -80,6 +90,19 @@
 	&.mce-content-body {
 		line-height: $editor-line-height;
 	}
+
+	div.mce-resizehandle {
```
[the remark this answers is not in the document]
Comment (iseulde) [reply editor/components/rich-text/style.scss]: Remnant from #5794 where it was designed like this intentionally. I'm don't really have a strong opinion here. @jasmussen what do you think?

File: editor/components/rich-text/tokens/registration.js [hunk 13]
```diff
@@ -0,0 +1,150 @@
...
+ * WordPress dependencies
+ */
+import { applyFilters } from '@wordpress/hooks';
+
+/**
+ * Browser dependencies
+ */
+const { error } = window.console;
+
+const tokenSettings = {};
```
[the remark this answers is not in the document]
Comment (iseulde) [reply editor/components/rich-text/tokens/registration.js]: Sure, I can change this. Could you enlighten me on the benefits of that approach?

[the remark this answers is not in the document]
Comment (youknowriad) [reply editor/components/rich-text/tokens/registration.js]: The main benefit is that `withSelect` is reactive. If we change the inner tokens (or register new ones) at runtime, the components relying on them will update.

Consistency is also an argument for this. Handle state across the application in a unique way. which is also extensible.

File: editor/components/rich-text/style.scss (hunk 12 above)
[the remark this answers is not in the document]
Comment (jasmussen) [reply editor/components/rich-text/style.scss]: Last I looked at the inline image branch, the resize handles were those same blue dots we use for regular image blocks. This would be ideal, IMO. Haven't had time to look at this branch, but purely from the discussion, consolidation in design would be 👍 👍

**Comment by iseulde** [comment]
Cool, then I’ll change it to theme colors. Not sure why white was used.

File: editor/components/rich-text/style.scss (hunk 12 above)
[the remark this answers is not in the document]
Comment (jayshenk) [reply editor/components/rich-text/style.scss]: Originally I had made the resize handles the same blue design as the image block, and then changed to a white background after this comment from @mtias https://github.com/WordPress/gutenberg/pull/5794#discussion_r185719064. I don't have a strong opinion either way though.

File: editor/components/inserter/index.js [hunk 14]
```diff
@@ -22,12 +24,6 @@ class Inserter extends Component {
 	onToggle( isOpen ) {
 		const { onToggle } = this.props;
 
-		if ( isOpen ) {
-			this.props.showInsertionPoint();
-		} else {
-			this.props.hideInsertionPoint();
-		}
```
[the remark this answers is not in the document]
Comment (iseulde) [reply editor/components/inserter/index.js]: Fixed

File: editor/components/rich-text/tokens/registration.js (hunk 13 above)
[the remark this answers is not in the document]
Comment (iseulde) [reply editor/components/rich-text/tokens/registration.js]: Sound good, I'll attempt to update it.

**Comment by mtias** [comment]
I wanted the handlers to be different because the image behaviour is different. Having same handlers would be another extra incentive for people to expect other alignment controls to be the same. cc @jasmussen

**Comment by jasmussen** [comment]
> I wanted the handlers to be different because the image behaviour is different. Having same handlers would be another extra incentive for people to expect other alignment controls to be the same. 

That's a good argument. I don't know that white handlers will work, though. Can we make them small and square? 

So big and round for Block level images, and small and square for Inline level images. Same blue color, same white border for contrast.

**Comment by iseulde** [comment]
@youknowriad @mtias I changed it so the tokens are registered in the `core/editor` store, but the registration feel a bit clunky. Was thinking that maybe each instance should have a store, but then the registered tokens never change across instances. Also I'm getting undefined error at `withDispatch` in the test suite that I'm not getting otherwise. Unsure why.

File: editor/components/rich-text/index.js [hunk 15]
```diff
@@ -130,6 +133,14 @@ export class RichText extends Component {
 		};
 
 		this.containerRef = createRef();
+
+		if ( ! coreTokensRegistered ) {
+			forOwn( tokens, ( { name, settings } ) => {
+				registerToken( name, settings );
```
Comment (youknowriad) [inline editor/components/rich-text/index.js]: I think with the way `withDispatch` work, it's not certain that `registerToken` is available in the constructor. Maybe more in `componentDidMount`

File: editor/store/reducer.js [hunk 16]
```diff
@@ -1091,6 +1099,99 @@ export function autosave( state = null, action ) {
...
+ *
+ * @return {Object} Updated state.
+ */
+export function tokens( state = {}, action ) {
+	switch ( action.type ) {
+		case 'REGISTER_TOKEN': {
+			const { name } = action;
+
+			if ( typeof name !== 'string' ) {
+				error(
```
Comment (youknowriad) [inline editor/store/reducer.js]: Interesting that you chose to do the validation in the reducers. I think in redux's theory, the reducers should be pure (free of side effects) to be able to run them multiple times, rewind etc... without side effects. I wonder if we should move the validation into the `registerToken` action creator instead (by allowing `null` being returned from action creators). cc @aduth

File: editor/components/rich-text/index.js (hunk 15 above)
Comment (iseulde) [reply editor/components/rich-text/index.js]: Why is that so? Why only so late in `componentDidMount`?

Comment (youknowriad) [reply editor/components/rich-text/index.js]: I'm not certain about this assumption honestly :). I was thinking the proxy methods created by `withDispatch` will be available later but looking at the code now, it seems that I was wrong. So nvm this one

Comment (iseulde) [reply editor/components/rich-text/index.js]: Hm, I'm not understanding why it's not working in the tests specifically... Not sure what I'm missing here. :/

File: editor/store/reducer.js (hunk 16 above)
Comment (iseulde) [reply editor/store/reducer.js]: So should I move the validation to the action creators, or would it not belong there either? Should it me moved back to separate functions? Do the token settings even belong in the `core/editor` store at all?

Comment (youknowriad) [reply editor/store/reducer.js]: I think action creators is the best place for it but I'm not certain the data module is ready to handle "empty actions" but should be easy to update to avoid calling `dispatch` on `null` actions.

Comment (iseulde) [reply editor/store/reducer.js]: Okay, let's keep a wrapper function for now.

**Comment by iseulde** [comment]
So I'm kind of blocked here with the question (1) where to store the data, if `core/editor` is there right place, and if so (2) how it should be implemented properly to only dispatch once for all `RichText` instances.

Maybe we should revert to storing to locally again and figure this out later?

**Comment by youknowriad** [comment]
> (1) where to store the data, if core/editor is there right place

Sounds like a good place for me (because it's a RichText specific API and RichText is in the editor module).

> (2) how it should be implemented properly to only dispatch once for all RichText instances.

Can't we just register those instances at init (once we create the store), it seems to me that these are just default tokens like any default redux state?

**Comment by iseulde** [comment]
> Can't we just register those instances at init (once we create the store), it seems to me that these are just default tokens like any default redux state?

I'm not quite following. So we'll make a separate function for validation. Then the token settings will need to run through there and somehow end up as the initial state of the redux store?

**Comment by youknowriad** [comment]
@iseulde I may be missing some context but I was just thinking of moving this:

```js
forOwn( tokens, ( { name, settings } ) => {
  store.dispatch( registerToken( name, settings ) );
} )
```

to the end of `editor/store/index.js`

**Comment by iseulde** [comment]
@youknowriad Okay I changed it like that. Not sure if it's entirely what you mean though. I feel like the token validation should somehow be a part of the store as it depends on the state, but it works like this too.

**Comment by iseulde** [comment]
The test errors, unsure why. Again it passes locally.

```
FAIL test/e2e/specs/adding-inline-tokens.test.js (34.279s)
  adding inline tokens
    ✕ Should insert inline image (31267ms)
  ● adding inline tokens › Should insert inline image
    waiting failed: timeout 30000ms exceeded
      
      at WaitTask._timeoutTimer.setTimeout (node_modules/puppeteer/lib/FrameManager.js:793:58)
      at Timeout.callback (node_modules/jsdom/lib/jsdom/browser/Window.js:628:19)
```

**Comment by grappler** [comment]
I tested this out today as I am interested to implement my own inline element.

Currently to be able to add an inline image, you need to add the inline image and then the text can be added around it afterwards. Though I find the natural workflow for me is to write my text first and then find an image that would fit in.

I sort of expected a media button like with the link icon in the toolbar.

In the next phase it would be cool to be able to choose the image size.

**Comment by youknowriad** [comment]
Can't reproduce the test breakage locally neither 🤔

**Comment by iseulde** [comment]
Restarted the job but it still fails. Unsure at this moment what I can do to debug it. Before the last (few) rebases, it also worked on Travis.

**Comment by youknowriad** [comment]
@iseulde Maybe we can add `console.log` or something between each step in the test to at least know where it fails?

**Comment by mcsf** [comment]
Noting that I do see the test failing locally.

```
$ ./bin/run-e2e-tests.sh
…
 FAIL  test/e2e/specs/adding-inline-tokens.test.js (35.362s)
  adding inline tokens
    ✕ Should insert inline image (31565ms)

  ● adding inline tokens › Should insert inline image

    waiting failed: timeout 30000ms exceeded

      at WaitTask._timeoutTimer.setTimeout (node_modules/puppeteer/lib/FrameManager.js:793:58)
      at Timeout.callback (node_modules/jsdom/lib/jsdom/browser/Window.js:628:19)

Test Suites: 1 failed, 17 passed, 18 total
```

**Comment by mcsf** [comment]
Interestingly, here's a different error now:

```
$ npm run test-e2e
…
 FAIL  test/e2e/specs/adding-inline-tokens.test.js (10.144s)
  adding inline tokens
    ✕ Should insert inline image (976ms)

  ● adding inline tokens › Should insert inline image

    TypeError: Cannot read property 'click' of undefined

       97 |     await page.click( '.edit-post-more-menu [aria-label="More"]' );
       98 |     const [ button ] = await page.$x( `//button[contains(text(), \'${ mode } Editor\')]` );
    >  99 |     await button.click( 'button' );
      100 | }
      101 |
      102 | export async function getHTMLFromCodeEditor() {

      at _callee6$ (test/e2e/support/utils.js:99:15)
      at tryCatch (node_modules/regenerator-runtime/runtime.js:62:40)
      at Generator.invoke [as _invoke] (node_modules/regenerator-runtime/runtime.js:296:22)
      at Generator.prototype.(anonymous function) [as next] (node_modules/regenerator-runtime/runtime.js:114:21)
      at step (test/e2e/support/utils.js:455:191)
      at test/e2e/support/utils.js:455:361
```

**Comment by iseulde** [comment]
@mcsf did you supply username, password, and/or base URL if it's different?

**Comment by iseulde** [comment]
So the problem seems to appear at `await page.waitForSelector( '.media-modal li.attachment' );`...

**Comment by iseulde** [comment]
So my tests pass locally headless and with-head. Travis only has this error with the media modal, unsure what's causing it. The first time selecting in the media modal works, the second time it does not.

**Review by youknowriad — APPROVED** [review]
Let's get this in 👍

**Comment by iseulde** [comment]
Yay

**Comment by jasmussen** [comment]
H O L Y  G U A C A M O L E

🎉

**Comment by mtias** [comment]
@iseulde thanks! :)

File: test/e2e/specs/adding-inline-tokens.test.js [hunk 17]
```diff
@@ -0,0 +1,47 @@
...
+		await inputElement.uploadFile( tmpFileName );
+
+		// Wait for upload.
+		await page.waitForSelector( '.media-modal li[aria-label="' + filename + '"]' );
+
+		// Insert the uploaded image.
+		await page.click( '.media-modal button.media-button-select' );
+
+		// Check the content.
+		const regex = new RegExp( '<!-- wp:paragraph -->\\s*<p>a\\u00A0<img class="wp-image-\\d+" style="width:10px" src="[^"]+\\/' + filename + '\\.png" alt="" \\/><\\/p>\\s*<!-- \\/wp:paragraph -->' );
```
Comment (aduth) [inline test/e2e/specs/adding-inline-tokens.test.js]: Do we really expect the `\u00A0` character here? Why not just a plain space?

**Comment by hegerpkd** [comment]
Is there a way I can de-register inline blocks? It is not apparent from the code to me. I would like to have the ability to switch off the inline images block.

I can use unregisterBlockType for blocks, but not for components and inline image.

**Comment by aduth** [comment]
Where is the developer documentation reference for inline blocks?

**Comment by iseulde** [comment]
@hegerpkd @aduth After #7890, I'd like to merge some of the current `RichText` extension points into one API. I'd prefer to leave current documentation as is until Gutenberg itself uses it fully for it's own `RichText` extension such as mentions.

**Comment by tomislavp83** [comment]
@hegerpkd any updates on documentation ? I would really like to know how to register my custom inline block (token). I had some luck with registerToken function but without documentation it's usless.

**Comment by designsimply** [comment]
@tomislavp83 comments on closed pull requests sometimes won't get seen easily. Would you mind opening a new issue for that?
